File: destack-py/destack/store/memory/edit.py
from collections import OrderedDict
from collections.abc import Sequence
from typing import assert_never

import structlog
from opentelemetry import trace

from destack.language import (
    CASCADING_EDIT_TYPES,
    EdgeDirection,
    EditEvent,
    EditOperation,
    EditType,
    IsArchivable,
    IsDeletable,
    Node,
    NodeDefinitionReference,
    NodeReference,
    NodeType,
)

from .core import MemoryContext, VersionedNodeKey
from .wiring import pack_node_row

tracer = trace.get_tracer(__name__)
logger = structlog.get_logger(__name__)

NODE_PARENT_KEY = str(Node.property("parent").id)

ARCHIVED_AT_KEY = str(IsArchivable.property("archived_at").id)
DELETED_AT_KEY = str(IsDeletable.property("deleted_at").id)


@tracer.start_as_current_span("memory.execute_edits")
def execute_edits(
    context: MemoryContext, events: Sequence[EditEvent]
) -> tuple[Sequence[EditEvent], Sequence[EditEvent]]:
    """Execute the Events."""
    assert events, f"no Events in {events!r}"

    edits = _optimize_edits(context, events)
    cascaded_edits: list[EditEvent] = []
    applied_edits: list[EditEvent] = []

    current_definition = NodeDefinitionReference.of(events[0].node_ptr)
    current_edit_type = events[0].type
    current_batch: list[EditEvent] = []
    for edit in edits:
        edit_definition = NodeDefinitionReference.of(edit.node_ptr)
        if (
            edit_definition.node_type != current_definition.node_type
            or edit.type != current_edit_type
        ):
            batch_applied_edits, batch_cascaded_edits = _execute_data_edit(
                context=context,
                definition=current_definition,
                edit_type=current_edit_type,
                edits=current_batch,
            )
            current_definition = edit_definition
            current_edit_type = edit.type
            current_batch = []

        current_batch.append(edit)

    if current_batch:
        batch_applied_edits, batch_cascaded_edits = _execute_data_edit(
            context=context,
            definition=current_definition,
            edit_type=current_edit_type,
            edits=current_batch,
        )
        applied_edits.extend(batch_applied_edits)
        cascaded_edits.extend(batch_cascaded_edits)

    logger.trace("memory.execute_edits", events=len(events), span="current")
    return applied_edits, cascaded_edits


@tracer.start_as_current_span("memory.optimize_edits")
def _optimize_edits(context: MemoryContext, edits: Sequence[EditEvent]) -> list[EditEvent]:
    """
    Optimize the Change/Edits *while retaining semantic equivalence*.
    Reorder and batch non-interfering Edits to minimize roundtrips.
    """

    optimized_edits: list[EditEvent] = []
    buffer: list[EditEvent] = []

    def flush():
        if not buffer:
            return
        grouped: OrderedDict[tuple[NodeType, EditType], list[EditEvent]] = OrderedDict()
        for e in buffer:
            key = (e.node_ptr.type, e.type)
            if key not in grouped:
                grouped[key] = []
            grouped[key].append(e)  # contiguously batched
        for batch in grouped.values():
            optimized_edits.extend(batch)  # contiguously batched
        buffer.clear()

    for edit in edits:
        if edit.type in CASCADING_EDIT_TYPES:
            flush()  # close current segment
            optimized_edits.append(edit)  # keep position
        else:
            buffer.append(edit)  # postpone

    flush()  # trailing segment
    return optimized_edits


@tracer.start_as_current_span("memory.execute_cascade")
def _execute_cascade(
    context: MemoryContext,
    definition: NodeDefinitionReference,
    node_ptrs: Sequence[NodeReference],
) -> Sequence[NodeReference]:
    """Get the cascaded Nodes for an Edit."""
    from .query import _walk_node

    child_ptrs = _walk_node(
        context=context,
        definition=definition,
        roots_ptr=node_ptrs,
        roots_parents_ptr=(),
        direction=EdgeDirection.CHILD,
        depth=1,
        where=None,
        snapshot_path=(),
    )
    return child_ptrs


@tracer.start_as_current_span("memory.execute_data_edit")
def _execute_data_edit(
    context: MemoryContext,
    definition: NodeDefinitionReference,
    edit_type: EditType,
    edits: Sequence[EditEvent],
) -> tuple[Sequence[EditEvent], Sequence[EditEvent]]:
    """
    Execute the Edits to the data (data only, no schema).
    Returns the applied Edits and any cascaded Edits.
    """
    from destack.language import ScalarType

    table = context.get(definition)

    # create/upsert
    if edit_type == EditType.CREATE or edit_type == EditType.UPSERT:
        for edit in edits:
            assert edit.value is not None, f"no value for {edit!r}"
            snapshot_id = edit.snapshot_ptr.id if edit.snapshot_ptr is not None else None
            node_key = VersionedNodeKey(id=edit.node_ptr.id, snapshot_id=snapshot_id)
            if edit_type == EditType.UPSERT or node_key not in table.rows:
                row = pack_node_row(table, edit.value)
                table.rows[node_key] = row
                table.rows_by_snapshot[snapshot_id][node_key.id] = row
                if row.parent_ptr is not None:
                    parent_table = context.get(row.parent_ptr)
                    parent_key = VersionedNodeKey(id=row.parent_ptr.id, snapshot_id=snapshot_id)
                    parent_table.rows_by_parent[parent_key].append(row)

        logger.trace(
            f"memory.{edit_type.name.lower()}",
            edits=len(edits),
            span="current",
        )
        return edits, ()

    # update
    elif edit_type == EditType.UPDATE:
        for edit in edits:
            snapshot_id = edit.snapshot_ptr.id if edit.snapshot_ptr is not None else None
            assert edit.attribute is not None, f"no prop_ptr for {edit!r}"
            prop = edit.attribute.resolve()
            assert prop is not None, f"no prop for {edit!r}"
            node_key = VersionedNodeKey(id=edit.node_ptr.id, snapshot_id=snapshot_id)
            if row := table.rows.get(node_key):
                if edit.operation == EditOperation.SET:
                    assert edit.value is not None, f"no value for {edit!r}"
                    row.value[str(prop.id)] = edit.value.value
                elif edit.operation == EditOperation.CLEAR:
                    row.value.pop(str(prop.id), None)
                else:
                    raise RuntimeError(f"unsupported operation: {edit!r}")

        logger.trace(
            f"memory.{edit_type.name.lower()}",
            edits=len(edits),
            span="current",
        )
        return edits, ()

    # move
    elif edit_type == EditType.MOVE:
        for edit in edits:
            assert edit.value is not None and edit.value.value is not None, f"no value for {edit!r}"
            assert edit.value.type.scalar_type == ScalarType.NODE_REFERENCE, (
                f"unexpected value: {edit!r}"
            )
            snapshot_id = edit.snapshot_ptr.id if edit.snapshot_ptr is not None else None
            node_key = VersionedNodeKey(id=edit.node_ptr.id, snapshot_id=snapshot_id)
            if row := table.rows.get(node_key):
                # remove from old parent
                if row.parent_ptr is not None:
                    parent_table = context.get(row.parent_ptr)
                    parent_key = VersionedNodeKey(id=row.parent_ptr.id, snapshot_id=snapshot_id)
                    parent_table.rows_by_parent[parent_key].remove(row)
                # update parent pointer
                row.parent_ptr = NodeReference.from_value(edit.value.value)
                row.value[NODE_PARENT_KEY] = edit.value.value
                # add to new parent
                if row.parent_ptr is not None:
                    parent_table = context.get(row.parent_ptr)
                    parent_key = VersionedNodeKey(id=row.parent_ptr.id, snapshot_id=snapshot_id)
                    parent_table.rows_by_parent[parent_key].append(row)

        logger.trace(
            f"memory.{edit_type.name.lower()}",
            edits=len(edits),
            span="current",
        )
        return edits, ()

    # archive/unarchive/delete/restore
    elif edit_type in (
        EditType.ARCHIVE,
        EditType.UNARCHIVE,
        EditType.DELETE,
        EditType.RESTORE,
    ):
        # cascade
        cascaded_node_ptrs = _execute_cascade(
            context=context,
            definition=definition,
            node_ptrs=tuple(edit.node_ptr for edit in edits),
        )
        cascaded_edits = tuple(
            EditEvent(type=edit_type, node_ptr=node_ptr) for node_ptr in cascaded_node_ptrs
        )

        # update timestamps
        for node_ptr in cascaded_node_ptrs:
            node_table = context.get(node_ptr)
            snapshot_id = node_ptr.snapshot_id if node_ptr.snapshot_id is not None else None
            node_key = VersionedNodeKey(id=node_ptr.id, snapshot_id=snapshot_id)
            if row := node_table.rows.get(node_key):
                if edit_type == EditType.ARCHIVE:
                    row.value[ARCHIVED_AT_KEY] = edits[0].created_at
                elif edit_type == EditType.UNARCHIVE:
                    row.value.pop(ARCHIVED_AT_KEY, None)
                elif edit_type == EditType.DELETE:
                    row.value[DELETED_AT_KEY] = edits[0].created_at
                elif edit_type == EditType.RESTORE:
                    row.value.pop(DELETED_AT_KEY, None)
                else:
                    assert_never(edit_type)

        logger.trace(
            f"memory.{edit_type.name.lower()}",
            edits=len(edits),
            span="current",
        )
        return edits, cascaded_edits

    # erase
    elif edit_type == EditType.ERASE:
        # cascade
        cascaded_node_ptrs = _execute_cascade(
            context=context,
            definition=definition,
            node_ptrs=tuple(edit.node_ptr for edit in edits),
        )
        cascaded_edits = tuple(
            EditEvent(type=edit_type, node_ptr=node_ptr) for node_ptr in cascaded_node_ptrs
        )

        # delete rows
        for node_ptr in cascaded_node_ptrs:
            node_table = context.get(node_ptr)
            snapshot_id = node_ptr.snapshot_id if node_ptr.snapshot_id is not None else None
            node_key = VersionedNodeKey(id=node_ptr.id, snapshot_id=snapshot_id)
            row = node_table.rows.pop(node_key, None)
            if row is not None:
                node_table.rows_by_snapshot[snapshot_id].pop(node_key.id, None)
                if not node_table.rows_by_snapshot[snapshot_id]:
                    node_table.rows_by_snapshot.pop(snapshot_id, None)
                if row.parent_ptr is not None:
                    parent_table = context.get(row.parent_ptr)
                    parent_key = VersionedNodeKey(id=row.parent_ptr.id, snapshot_id=snapshot_id)
                    parent_table.rows_by_parent[parent_key].remove(row)

        logger.trace(
            f"memory.{edit_type.name.lower()}",
            edits=len(edits),
            span="current",
        )
        return edits, cascaded_edits

    else:
        assert_never(edit_type)

File: destack-ts/src/store/memory/entity/edit.ts
import {
  CASCADING_EDIT_TYPES,
  EdgeDirection,
  EditEvent,
  EditOperation,
  EditType,
  NodeDefinitionReference,
  NodeReference,
  ScalarType,
} from "@destack/language";
import {
  NODE_DELETED_AT_KEY as ENTITY_DELETED_AT_KEY,
  ENTITY_PARENT_KEY,
  MAX_RECURSION_DEPTH,
  MemoryContext,
} from "@destack/store/memory/core";
import { walkNode } from "@destack/store/memory/entity/query";
import { packEntityRow } from "@destack/store/memory/entity/wiring";
import { getLogger, getTracer, traceFunction } from "@destack/utils";

const logger = getLogger("memory.entity.edit");
const tracer = getTracer("memory.entity.edit");

/**
 * Execute the Edits in-memory.
 */
function _executeEdits(options: { context: MemoryContext; edits: EditEvent[] }): {
  edits: EditEvent[];
  cascadedEdits: EditEvent[];
} {
  const { context, edits } = options;
  if (edits.length === 0) {
    return { edits: [], cascadedEdits: [] };
  }

  const optimizedEdits = optimizeEdits({ context, edits });
  const cascadedEdits: EditEvent[] = [];
  const appliedEdits: EditEvent[] = [];

  let currentDefinition = NodeDefinitionReference.of(optimizedEdits[0].nodePtr);
  let currentEditType = optimizedEdits[0].type;
  let currentBatch: EditEvent[] = [];

  for (const edit of edits) {
    const editDefinition = NodeDefinitionReference.of(edit.nodePtr);
    if (editDefinition.nodeType !== currentDefinition.nodeType || edit.type !== currentEditType) {
      const { edits: batchAppliedEdits, cascadedEdits: batchCascadedEdits } = executeEdit({
        context,
        edits: currentBatch,
        definition: currentDefinition,
        editType: currentEditType,
      });
      appliedEdits.push(...batchAppliedEdits);
      cascadedEdits.push(...batchCascadedEdits);
      currentDefinition = editDefinition;
      currentEditType = edit.type;
      currentBatch = [];
    }
    currentBatch.push(edit);
  }

  // flush last batch
  if (currentBatch.length > 0) {
    const { edits: batchAppliedEdits, cascadedEdits: batchCascadedEdits } = executeEdit({
      context,
      edits: currentBatch,
      definition: currentDefinition,
      editType: currentEditType,
    });
    appliedEdits.push(...batchAppliedEdits);
    cascadedEdits.push(...batchCascadedEdits);
  }

  return { edits: appliedEdits, cascadedEdits };
}
export const executeEdits = traceFunction(tracer, "execute_edits", _executeEdits);

/**
 * Optimize the Edits while retaining semantic equivalence.
 * Reorder and batch non-interfering Edits to minimize roundtrips.
 */
function optimizeEdits(options: { context: MemoryContext; edits: EditEvent[] }): EditEvent[] {
  const { context, edits } = options;
  const optimizedEdits: EditEvent[] = [];
  const buffer: EditEvent[] = [];

  function flush() {
    if (buffer.length === 0) {
      return;
    }
    const grouped = new Map<string, EditEvent[]>();
    for (const e of buffer) {
      const key = `${e.nodePtr.type}:${e.type}`;
      if (!grouped.has(key)) {
        grouped.set(key, []);
      }
      grouped.get(key)!.push(e);
    }
    for (const batch of grouped.values()) {
      optimizedEdits.push(...batch);
    }
    buffer.length = 0;
  }

  for (const edit of edits) {
    if (CASCADING_EDIT_TYPES.includes(edit.type)) {
      flush(); // close current segment
      optimizedEdits.push(edit); // keep position
    } else {
      buffer.push(edit); // postpone
    }
  }

  flush(); // trailing segment
  return optimizedEdits;
}

/**
 * Get the cascaded Nodes for an Edit.
 */
function _executeCascade(options: {
  context: MemoryContext;
  definition: NodeDefinitionReference;
  nodePtrs: NodeReference[];
  includeDeleted?: boolean | Array<string>;
}): { cascadedNodePtrs: NodeReference[]; sourceIdByNodeId: Map<string, string> } {
  const { context, definition, nodePtrs, includeDeleted } = options;
  const { cascadedNodePtrs, sourceIdByNodeId } = walkNode({
    context,
    definition,
    nodesPtrs: nodePtrs,
    direction: EdgeDirection.CHILD,
    depth: MAX_RECURSION_DEPTH,
    includeDeleted,
    snapshotPath: [],
  });
  return { cascadedNodePtrs, sourceIdByNodeId };
}
const executeCascade = traceFunction(tracer, "execute_cascade", _executeCascade);

/**
 * Execute the Edits to the data (data only, no schema).
 * Returns the applied Edits and any cascaded Edits.
 */
function _executeEdit(options: {
  context: MemoryContext;
  definition: NodeDefinitionReference;
  editType: EditType;
  edits: EditEvent[];
}): { edits: EditEvent[]; cascadedEdits: EditEvent[] } {
  const { context, definition, edits, editType } = options;
  const table = context.getEntityTable(definition);

  // create/upsert
  if (editType === EditType.CREATE || editType === EditType.UPSERT) {
    for (const edit of edits) {
      if (!edit.value) {
        throw new Error(`no value for ${edit.repr()}`);
      }
      const snapshotId = edit.snapshotPtr ? edit.snapshotPtr.id : null;
      const nodeKey = table.getNodeKey({ id: edit.nodePtr.id, snapshotId });
      if (editType === EditType.UPSERT || !table.rows.has(nodeKey)) {
        const row = packEntityRow(edit.value);
        // main table
        table.rows.set(nodeKey, row);
        // rowsBySnapshot
        if (!table.rowsBySnapshot.has(snapshotId)) {
          table.rowsBySnapshot.set(snapshotId, new Map());
        }
        table.rowsBySnapshot.get(snapshotId)!.set(edit.nodePtr.id, row);
        // parent-child relationships
        if (row.parentPtr) {
          const parentTable = context.getEntityTable(row.parentPtr);
          const parentKey = parentTable.getNodeKey({ id: row.parentPtr.id, snapshotId });
          if (!parentTable.rowsByParent.has(parentKey)) {
            parentTable.rowsByParent.set(parentKey, []);
          }
          parentTable.rowsByParent.get(parentKey)!.push(row);
        }
      }
    }
    return { edits, cascadedEdits: [] };
  }

  // update
  else if (editType === EditType.UPDATE) {
    for (const edit of edits) {
      const snapshotId = edit.snapshotPtr ? edit.snapshotPtr.id : null;
      if (!edit.propertyId) {
        throw new Error(`no propertyId for ${edit.repr()}`);
      }
      const nodeKey = table.getNodeKey({ id: edit.nodePtr.id, snapshotId });
      const row = table.rows.get(nodeKey);
      if (!row) {
        throw new Error(`node not found for ${edit.repr()}: ${edit.nodePtr.repr()}`);
      } else if (edit.operation === EditOperation.SET) {
        if (!edit.value) {
          throw new Error(`no value for ${edit.repr()}`);
        }
        row.value[String(edit.propertyId)] = edit.value.value;
      } else {
        throw new Error(`unsupported operation: ${edit.repr()}`);
      }
    }
    return { edits, cascadedEdits: [] };
  }

  // move
  else if (editType === EditType.MOVE) {
    for (const edit of edits) {
      if (!edit.value) {
        throw new Error(`no value for ${edit.repr()}`);
      }
      if (edit.value.type.scalarType !== ScalarType.NODE_REFERENCE) {
        throw new Error(`unexpected value: ${edit.repr()}`);
      }
      const snapshotId = edit.snapshotPtr ? edit.snapshotPtr.id : null;
      const nodeKey = table.getNodeKey({ id: edit.nodePtr.id, snapshotId });
      const row = table.rows.get(nodeKey);
      if (!row) {
        throw new Error(`node not found for ${edit.repr()}: ${edit.nodePtr.repr()}`);
      }
      // remove from old parent
      if (row.parentPtr) {
        const parentTable = context.getEntityTable(row.parentPtr);
        const parentKey = parentTable.getNodeKey({ id: row.parentPtr.id, snapshotId });
        const children = parentTable.rowsByParent.get(parentKey);
        if (children) {
          const index = children.indexOf(row);
          if (index >= 0) {
            children.splice(index, 1);
          }
        }
      }
      // update parent pointer
      row.parentPtr = NodeReference.fromValue(edit.value.value);
      row.value[ENTITY_PARENT_KEY] = edit.value.value;
      // add to new parent
      if (row.parentPtr) {
        const parentTable = context.getEntityTable(row.parentPtr);
        const parentKey = parentTable.getNodeKey({ id: row.parentPtr.id, snapshotId });
        if (!parentTable.rowsByParent.has(parentKey)) {
          parentTable.rowsByParent.set(parentKey, []);
        }
        parentTable.rowsByParent.get(parentKey)!.push(row);
      }
    }
    return { edits, cascadedEdits: [] };
  }

  // delete/restore
  else if (editType === EditType.DELETE || editType === EditType.RESTORE) {
    // cascade
    const nodesPtrs = edits.map((edit) => edit.nodePtr);
    const editByNodeId = new Map<string, EditEvent>();
    for (const edit of edits) {
      editByNodeId.set(edit.nodePtr.id, edit);
    }

    let includeDeleted: boolean | Array<string> = false;
    if (editType === EditType.RESTORE) {
      // restrict to nodes with same deleted_at
      includeDeleted = [];
      for (const nodePtr of nodesPtrs) {
        const snapshotId = nodePtr.snapshotId || null;
        const nodeKey = table.getNodeKey({ id: nodePtr.id, snapshotId });
        const row = table.rows.get(nodeKey);
        if (!row) {
          throw new Error(`node not found: ${nodePtr.repr()}`);
        } else {
          const deletedAt = row.value[ENTITY_DELETED_AT_KEY];
          if (deletedAt && !includeDeleted.includes(deletedAt)) {
            includeDeleted.push(deletedAt);
          }
        }
      }
    }

    const { cascadedNodePtrs, sourceIdByNodeId } = executeCascade({
      context,
      definition,
      nodePtrs: nodesPtrs,
      includeDeleted,
    });
    const cascadedEdits = cascadedNodePtrs.map(
      (nodePtr) => new EditEvent({ type: editType, node: nodePtr }),
    );

    // update timestamps
    for (const nodePtr of [...nodesPtrs, ...cascadedNodePtrs]) {
      const edit = editByNodeId.get(sourceIdByNodeId.get(nodePtr.id) || nodePtr.id);
      if (!edit) {
        throw new Error(`edit not found for ${nodePtr.repr()}`);
      }
      const snapshotId = edit.snapshotPtr?.id || null;
      const nodeKey = table.getNodeKey({ id: nodePtr.id, snapshotId });
      const row = table.rows.get(nodeKey);
      if (!row) {
        throw new Error(`node not found for ${edit.repr()}: ${nodePtr.repr()}`);
      } else if (editType === EditType.DELETE) {
        row.value[ENTITY_DELETED_AT_KEY] = edit.createdAt.toString({ timeZoneName: "never" });
      } else if (editType === EditType.RESTORE) {
        delete row.value[ENTITY_DELETED_AT_KEY];
      }
    }

    return { edits, cascadedEdits };
  }

  //
  else {
    throw new Error(`unsupported edit type: ${EditType[editType]}`);
  }
}
const executeEdit = traceFunction(tracer, "execute_edit", _executeEdit);
